src/utf8.rs

```diff
@@ -0,0 +1,693 @@
+    #[cfg(all(loom, test))]
+    pub fn from_static(string: &'static str) -> Self {
+        let bytes = Bytes::from_static(string.as_bytes());
+        // SAFETY: bytes was created from a UTF-8 string
+        unsafe { Self::from_utf8_unchecked(bytes) }
+    }
+
+    /// Convert `Bytes` to `Utf8Bytes`.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the argument is not valid UTF-8.
+    #[inline]
+    pub fn from_utf8(bytes: Bytes) -> Result<Self, FromUtf8Error> {
+        match core::str::from_utf8(&bytes) {
+            // SAFETY: verified that b is valid UTF-8
+            Ok(_) => Ok(unsafe { Utf8Bytes::from_utf8_unchecked(bytes) }),
+            Err(error) => Err(FromUtf8Error { bytes, error }),
+        }
+    }
+
+    /// Convert [`Bytes`] to [`Utf8Bytes`] without checking that it contains
+    /// valid UTF-8.
+    ///
+    /// # Safety
+    ///
+    /// The argument must contain valid UTF-8.
+    #[inline]
+    pub const unsafe fn from_utf8_unchecked(bytes: Bytes) -> Self {
+        Utf8Bytes { inner: bytes }
+    }
+
+    /// Unwrap `Utf8Bytes` into its underlying `Bytes`.
+    pub fn into_bytes(self) -> Bytes {
+        self.inner
+    }
+
+    /// Create [Utf8Bytes] with a buffer whose lifetime is controlled
+    /// via an explicit owner.
+    ///
+    /// See [`Bytes::from_owner`] for more details.
+    pub fn from_owner<T>(owner: T) -> Self
+    where
+        T: AsRef<str> + Send + 'static,
+    {
+        struct OwnerWrapper<T>(T);
+        impl<T: AsRef<str>> AsRef<[u8]> for OwnerWrapper<T> {
+            fn as_ref(&self) -> &[u8] {
+                self.0.as_ref().as_bytes()
+            }
+        }
+
+        let bytes = Bytes::from_owner(OwnerWrapper(owner));
+        // SAFETY: bytes was created from an owner whose AsRef<[u8]> impl always returns valid UTF-8
+        unsafe { Utf8Bytes::from_utf8_unchecked(bytes) }
+    }
+
+    /// Get a `&str` reference from this `Utf8Bytes`.
+    #[inline]
+    pub fn as_str(&self) -> &str {
+        // SAFETY: guaranteed sound by type-level invariant
+        unsafe { core::str::from_utf8_unchecked(&self.inner) }
+    }
+
+    /// Returns the number of bytes contained in this `Bytes`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use bytes::Utf8Bytes;
+    ///
+    /// let b = Utf8Bytes::from("hello");
+    /// assert_eq!(b.len(), 5);
+    /// ```
+    #[inline]
+    pub const fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    /// Returns true if the `Utf8Bytes` has a length of 0.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use bytes::Utf8Bytes;
+    ///
+    /// let b = Utf8Bytes::new();
+    /// assert!(b.is_empty());
+    /// ```
+    #[inline]
+    pub const fn is_empty(&self) -> bool {
+        self.inner.is_empty()
+    }
+
+    /// Creates `Utf8Bytes` instance from slice, by copying it.
+    pub fn copy_from_slice(data: &str) -> Self {
+        String::from(data).into()
```

Comment:
Should this call `Bytes::copy_from_slice` instead and avoid having more and more code rely on the internal representation of `Bytes`?

Reply:
Ah, good point. Does `copy_from_slice` seem like a reasonable name here? I figured that the `Bytes` version was probably moreso referring to the borrowed nature of `data`, rather than the fact it's specifically a `[_]` slice type, so I used the same name here instead of something like `copy_from_str`.